Optimize console output in ConsoleHost

iSazonov opened the pull request:
## PR Summary

- Implement `WriteLine` method to write to a host. Win is that we exclude extra string allocations like `value + Crlf ` and extra locks and kernel calls.
- Exclude extra allocation for every char in console input (`keyInfo.Character.ToString()`) in `ReadKey(ReadKeyOptions options)`.
- Exclude extra allocations in the console host by implementation of `WriteLine`-like methods to replace `Write(strValue + Crlf)` with `WriteLine(strValue)`.
- Remove unneeded locks because of replacing `Write(strValue); Write(Crlf)` with `WriteLine(strValue)`.
- Add new lock in `WriteToConsole(ConsoleColor foregroundColor, ConsoleColor backgroundColor, string text, bool newLine = false)`.
- Use CoreFX pattern for P/Invoke to support `ReadOnlySpan`.


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/powershell/6882)
<!-- Reviewable:end -->

iSazonov committed:
[Feature] Exclude extra allocations in ConsoleControl

iSazonov commented:
CI Appveyor temporary failed on Update-Help tests.

iSazonov commented:
Reopen the PR to restart CIs.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2613,18 +2613,29 @@ internal static void SetConsoleWindowTitle(string consoleTitle)
         ///
         /// </param>
         /// <exception cref="HostException">
+        /// <param name="newLine">
```
you put the param section in an exception section.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2613,18 +2613,29 @@ internal static void SetConsoleWindowTitle(string consoleTitle)
         /// </exception>
-
-        internal static void WriteConsole(ConsoleHandle consoleHandle, string output)
+        internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char> output, bool newLine)
```
Why adding the new parameter `newline`? Are you fixing a bug?

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -3090,15 +3117,30 @@ NakedWin32Handle templateFileWin32Handle
+                fixed (char* bufferPtr = &MemoryMarshal.GetReference(buffer))
+                {
+                    return WriteConsole(consoleOutput, bufferPtr, numberOfCharsToWrite, out numberOfCharsWritten, reserved);
+                }
```
This obviously complexes the code. Actually, all changes here make code more complex. Do we have measurable perf improvement to justify the changes? I don't see a large amount of string manipulation here (please point it out if I miss them). Note that many string APIs and `StringBuilder` already received perf improvements in 2.1.

iSazonov replied:
It looks like complement - I'm not sure I can create a _complex_ code :smile:

Exactly the pattern I found in CoreFX. This is due to C# language limitations.

Really the code isn't getting any more complicated. Rather on the contrary it becomes more complete.
By analyzing the code to clean up psl I found that in this place we make a lot of kernel calls and extra locks because of the multitude of:
```c#
WriteToConsole(text, true);
WriteToConsole(Crlf, true);
```
I was very surprised that we don't have a WriteLine() method to write to a host and use many allocations like `value + Crlf` - so I just implemented the method. 
The only complicity of the code is that I had to modify three levels of interfaces. So really the PR is very simple.

> Note that many string APIs and StringBuilder already received perf improvements in 2.1.

Yes, I saw up to 1100000 allocations in VS Profiler at pwsh.exe 6.0 startup and up to 850000 allocation at 6.1 (already on .Net 2.1) startup. Nevertheless, I still see a lot of allocations. The profile show that up to 36% allocations is chars and strings. 
This shows that CoreCLR does not solve all allocation problems and we should continue to consume resources cautiously and to optimize our code.

daxian-dbw replied:
I'm not convinced whether the startup is the right scenario to measure for changes in this PR, but still, for the startup scenario, what's the improvement in object allocation with your changes?

iSazonov replied:
No, it was only sample that 2.1 reduce allocations but not remove its at all. I don't know how measure wins in the PR - VS Profiler is crashed after first run. I look https://benchmarkdotnet.org/ but it takes a lot of effort. At least in the first steps.

iSazonov replied:
I attached a file with simple perf tests. I saw win ~2 second on 100000 outputs.
[measure_conhost.txt](https://github.com/PowerShell/PowerShell/files/2080705/measure_conhost.txt)

BrucePay replied:
Do we really need any of this code anymore? It was written back in V1 when the `Console` API was very limited. I suspect that we could remove most of the code with just a thin wrapper around the .NET APIs. (Which, in fact, were somewhat inspired by this code.)

iSazonov replied:
I agree that this code would have to be cleaned more thoroughly. The problem is that we do not have tests for the console code. Therefore, I would prefer to make small changes step by step so as not to break something. I intend to continue working in this direction (if only you do not chop it in one your PR :smile:).

iSazonov committed:
Fix param section

iSazonov commented:
@daxian-dbw I changed PR title - seems old one confuse you that is main win in the PR.

iSazonov committed:
Minor style corrections

iSazonov commented:
We could use https://benchmarkdotnet.org/ for perf analyses.

daxian-dbw commented:
@iSazonov I like the perf analysis idea. We definitely need it for benefits in the long run. I will look into it.

iSazonov commented:
@daxian-dbw Not related the PR - I passed 200 MB file through ConvertFrom-Csv and get alarming results.
On Windows PowerShell - 28 sec for every run.
On PowerShell Core 6.1 (crossgened) - from 33 to 55 sec! If I run and then wait some time - all results is ~33 sec. But if I run the test sequentially without pauses, the time starts with 33 and increase up to 55 sec. 
I have two fears. (1) .Net Core 2.1 has a problem with a caching. (2) why we are slower on PowerShell Core.

iSazonov commented:
@daxian-dbw @anmenaga @SteveL-MSFT Please review the PR.

iSazonov commented:
@daxian-dbw Could you please continue the PR review?

daxian-dbw commented:
@iSazonov I'm busy on some investigations and 6.1-must-have issues recently (RC will be next week and need to get those in). Will resume reviewing this PR in a few days.

daxian-dbw commented:
A ping for myself so that the stale bot doesn't kick in :)

iSazonov commented:
ping -l 1496 
:-)

iSazonov commented:
Rebased to update CIs.

iSazonov commented:
@SteveL-MSFT Have MSFT team special thoughts about the PR? Can we continue or close?

SteveL-MSFT commented:
@iSazonov @daxian-dbw is currently on vacation and will be back next week.

anmenaga commented:
@daxian-dbw friendly ping for review.

iSazonov commented:
This seems like it is the longest PR :-(
@daxian-dbw @SteveL-MSFT Could we continue?

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2520,36 +2532,55 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, string output)
-                    HostException e = CreateHostException(err, "WriteConsole",
-                        ErrorCategory.WriteError, ConsoleControlStrings.WriteConsoleExceptionTemplate);
-                    throw e;
+                    Span<char> outBufferLine = stackalloc char[outBuffer.Length + 2];
```
The `2` here is for `CrLf`?  Perhaps define a const to make this more readable

iSazonov replied:
It is different on Windows and Unix so we need use `Crlf.Length`.

Fixed.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2520,36 +2532,55 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, string output)
-                    throw e;
+                    Span<char> outBufferLine = stackalloc char[outBuffer.Length + 2];
+                    outBuffer.CopyTo(outBufferLine);
+                    ConsoleHostUserInterface.Crlf.AsSpan().CopyTo(outBufferLine.Slice(outBufferLine.Length-2));
```
Same here for `2` and also different whitespace for subtraction than addition above

iSazonov replied:
Fixed.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -581,34 +598,61 @@ internal void WriteToConsole(string value, bool transcribeResult)
+                {
+                    if (newLine)
+                    {
+                        WriteLineToConsole(text, true);
```
Should have named param for `true`

iSazonov replied:
Fixed.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -581,34 +598,61 @@ internal void WriteToConsole(string value, bool transcribeResult)
+                    }
+                    else
+                    {
+                        WriteToConsole(text, true);
```
Should have named param for `true`

iSazonov replied:
Fixed.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -710,16 +818,30 @@ public override void Write(ConsoleColor foregroundColor, ConsoleColor background
+        /// </exception>
+        public override void WriteLine()
+        {
+            this.WriteHelper("", newLine: true);
```
string.Empty

iSazonov replied:
Fixed.

iSazonov committed:
Fix doc comments

iSazonov committed:
Address feedback

iSazonov commented:
Rebased to get latest CI's updates.

iSazonov committed:
Style fix 1

iSazonov committed:
Style fix 2

iSazonov committed:
Style fix 3

iSazonov commented:
@SteveL-MSFT @TravisEz13 Should we now address Codacy issues?

SteveL-MSFT commented:
@iSazonov we only just enabled Codacy so I wouldn't consider it blocking right now.  Looking at the Codacy feedback for this PR, seems like good suggestions.

iSazonov commented:
@SteveL-MSFT Thanks! Now I'll ask contributors to look Codacy issues as optional.

iSazonov commented:
I did not find what I could fix from Codacy issues.

@daxian-dbw Can we merge?

iSazonov commented:
@daxian-dbw @SteveL-MSFT Can we merge before 6.2 Preview4?

SteveL-MSFT commented:
@iSazonov @daxian-dbw is currently on vacation, we can get this in for RC

iSazonov commented:
Can we merge or we are waiting @daxian-dbw ?

iSazonov commented:
@daxian-dbw Could we continue?

daxian-dbw commented:
@iSazonov Sorry for being tardy in my PR reviews recently (vacation and then was busy on another project). I will get back to this PR on Wednesday.

RDIL reviewed (commented):
@iSazonov I'm noticing conflicts.  Will need to be fixed now or later.

iSazonov commented:
Rebased to get latest environment updates.

>  was busy on another project

It is a pity that it is not visible and impossible to learn from you.

iSazonov commented:
@daxian-dbw Can we continue?

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2523,64 +2523,92 @@ internal static void SetConsoleWindowTitle(string consoleTitle)
-                    HostException e = CreateHostException(err, "WriteConsole",
-                        ErrorCategory.WriteError, ConsoleControlStrings.WriteConsoleExceptionTemplate);
-                    throw e;
+                    var endOfLine = ConsoleHostUserInterface.Crlf.AsSpan();
```
Can we change `ConsoleHostUserInterface.Crlf` to `Environment.NewLine`?

iSazonov replied:
It is defined as `internal static string Crlf = System.Environment.NewLine;`.
If we want to replace we should replace everywhere (13 points). Make sense?

daxian-dbw replied:
`Crlf` is a bad name given we are x-plats now. I'm fine that we replace everywhere.

iSazonov replied:
Will do in follow PR.

daxian-dbw reviewed (commented):
@iSazonov My apology for leaving this PR behind for soooo long.
Let's get this PR done next week, then we can see if it's possible to eliminate the native call to `WriteConsole` completely -- `Console.Write` should do the same thing.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleTextWriter.cs:
```diff
@@ -39,38 +39,61 @@ public override
         {
-            this.Write(new String(c, 1));
+            Span<char> c1 = stackalloc char[1];
+            c1[0] = c;
```
Done.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -541,40 +541,55 @@ private void BlankAtCursor(Coordinates cursorPosition)
+        internal void WriteToConsole(char c, bool transcribeResult)
+        {
+            Span<char> value = stackalloc char[1];
+            value[0] = c;
```
Done.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleTextWriter.cs:
```diff
@@ -39,38 +39,61 @@ public override
+            }
+            else
+            {
+                _ui.WriteToConsole(bool.FalseString.AsSpan(), transcribeResult: true);
```
Removed.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -716,16 +818,26 @@ public override void Write(ConsoleColor foregroundColor, ConsoleColor background
+        /// </exception>
+        public override void WriteLine()
+        {
+            this.WriteHelper(string.Empty, newLine: true);
```
I can not find the extra call.

daxian-dbw replied:
I mean when we move from the win32 API and call `Console.WriteLine` instead.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -587,34 +602,61 @@ internal void WriteToConsole(string value, bool transcribeResult)
+
+        private void WriteLineToConsole(ConsoleColor foregroundColor, ConsoleColor backgroundColor, string text)
+        {
+            WriteToConsole(PromptColor, backgroundColor, text, newLine: true);
```
It is typo.
Fixed.

iSazonov commented:
> My apology for leaving this PR behind for soooo long.

Almost a year I was tormented by guesswork :-)))

> Let's get this PR done next week, then we can see if it's possible to eliminate the native call to WriteConsole completely -- Console.Write should do the same thing.

In another PR we touched ReadConsole() canonical/cooked mode and that tab do not terminate input on Unix - it seems we need to consider it together.

daxian-dbw commented:
> In another PR we touched ReadConsole() canonical/cooked mode and that tab do not terminate input on Unix - it seems we need to consider it together.

Can you give the link to that PR? Or is it something you are still working on?

iSazonov commented:
https://github.com/PowerShell/PowerShell/pull/9165#discussion_r267153099

We have different code paths for Windows and Unix. `Colsole.Readline` does not support <kbd>Tab</kbd> cooked mode. I opened an Issue in CoreFX. Or we will have to migrate to `ReadKey()`. Also we could look at Bash Readline library.

iSazonov commented:
@daxian-dbw Is there something I need still to fix?

daxian-dbw commented:
@iSazonov Not at this time, but I'd like to go through the changes again before approve the PR.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -541,40 +541,49 @@ private void BlankAtCursor(Coordinates cursorPosition)
+                ConsoleOutWriteHelper(value, newLine);
             }
 
             if (transcribeResult)
```
> PostWrite(value);

This needs to be changed too, as a new line could be requested.

iSazonov replied:
Done.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -652,10 +706,22 @@ public override void Write(string value)
+
+                if (newLine)
+                {
+                    _parent.OutputSerializer.Serialize(Crlf);
```
Add the `Crlf` to value may be better than calling `Serialize` twice.

iSazonov replied:
Done.

daxian-dbw commented:
@iSazonov I pushed a commit to reorder `WriteLineToConsole` and make `WriteConsole` not call `'WriteLineToConsole'.

Overall, this PR avoids some string allocation (adding new line to text) with the cost of doubling the number of overload methods and adding more branches (if/else).
I don't know if this will improve performance slightly because more branches would hurt branch prediction.
From the maintenance point of view, doubling overloads with more branches make it harder to maintain the code.

So I'm not very sure that gains from this PR outweigh the costs ... @SteveL-MSFT what do you think?

iSazonov commented:
@daxian-dbw Thanks!

The current code reduces number of p/invokes, locks and allocations. All together gives a good result.
My main concern was extra allocations we do for every (!) new line. .Net Core GC is much more lazy than Framework GC. Using ReadOnlySpans greatly improves the process.

> the cost of doubling the number of overload methods and adding more branches (if/else).

I believe new code got much closer to Console.Write/WriteLine and we can more easily move to the methods.
I think we could add `MethodImplAttributes.AggressiveInlining` for some temporary methods. And it seems we can reduce number of branches - will do the commit.

iSazonov committed:
Address feedback

iSazonov committed:
Add AggressiveInlining

daxian-dbw commented:
@iSazonov Thanks for the further clarification.

> The current code reduces number of p/invokes, locks and allocations.

The P/Invoke will most likely go away by replacing it with `Console.Write/WriteLine`, but you are right that this PR will help when moving to `Console.Write/WriteLine`.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -542,80 +544,122 @@ private void BlankAtCursor(Coordinates cursorPosition)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteLineToConsole(ConsoleColor foregroundColor, ConsoleColor backgroundColor, string text)
+        {
+            WriteToConsole(PromptColor, backgroundColor, text, newLine: true);
```
You recent rebase wiped out the changes that addressed my comments. Please bring them back.

iSazonov replied:
It seems I mistakenly put a commit into my local master :-(
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2542,64 +2542,92 @@ internal static void SetConsoleWindowTitle(string consoleTitle)
-                        IntPtr.Zero);
-
-                if (result == false)
+                if (newLine)
```
Here you always add the new line even if the execution flow got in  the `if (cursor + MaxBufferSize < output.Length)` block, in which case no new line should be written out yet.

iSazonov replied:
Good catch! 
I still can't train myself to look wider :-(

Fixed.

iSazonov committed:
Fix broken rebase

iSazonov committed:
Put new line only after last position

daxian-dbw commented:
@iSazonov you still haven't brought back all previous changes that addressed my comments starting from https://github.com/PowerShell/PowerShell/pull/6882#discussion_r270650824
Please go over my comments again and address any missing ones.

iSazonov committed:
Fix init

iSazonov committed:
Update ConsoleTextWriter.cs

Remove AsSpan()

iSazonov committed:
Reorder methods

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -587,34 +602,61 @@ internal void WriteToConsole(string value, bool transcribeResult)
+            WriteToConsole(PromptColor, backgroundColor, text, newLine: true);
+        }
+
+        private void WriteToConsole(ConsoleColor foregroundColor, ConsoleColor backgroundColor, string text, bool newLine = false)
```
Done.

iSazonov commented:
@daxian-dbw Sorry for the inconvenience. I hope that everything is fixed

PoshChan commented:
@iSazonov, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

iSazonov commented:
@daxian-dbw I don't know that is wrong.  I added new commits and CIs failed. Reverted commits and CIs failed again. It seems something was changed in CIs?

daxian-dbw commented:
I think @adityapatwardhan made some changes so that now Linux/macOS CIs always run feature tests. It looks to me your changes cause some feature tests to fail.

iSazonov commented:
@daxian-dbw Thanks! I I looked through the CIs history - feature tests was turned on but not reported in right way then this was fixed and now we see the CIs fail.
I reviewed my changes and not found cause of the failure. I'm going to create a new git branch to find a problem step by step if you don't have a better idea.

PoshChan commented:
@iSazonov, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 414
414:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 430
430:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 474
474:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 394
394:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 394
394:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 414
414:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 430
430:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 474
474:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 394
394:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 394
394:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

iSazonov commented:
I discover that input echoing to output in the failed tests but I can not find the cause.

PoshChan commented:
@iSazonov, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

PoshChan commented:
@iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input w/ initial command
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   4
Strings differ at index 4.
Expected: 'PS> 1+1'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 413
413:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input explicit prompting (-File -)
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 429
429:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Redirected input w/ nested prompt
```powershell
Expected strings to be the same, but they were different.
Expected length: 29
Actual length:   4
Strings differ at index 4.
Expected: 'PS> $host.EnterNestedPrompt()'
But was:  'PS> '
---------------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 473
473:             $process.StandardOutput.ReadLine() | Should -Be "PS> `$host.EnterNestedPrompt()"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input: <empty>
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```
ConsoleHost unit tests.Redirected standard input for 'interactive' use.Interactive redirected input:  -IntERactive
```powershell
Expected strings to be the same, but they were different.
Expected length: 7
Actual length:   29
Strings differ at index 0.
Expected: 'PS> 1+1'
But was:  '$function:prompt = { 'PS> ' }'
-----------^
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 393
393:             $process.StandardOutput.ReadLine() | Should -Be "PS> 1+1"
```

iSazonov committed:
Fix Serialize

daxian-dbw committed:
Address my own comments

daxian-dbw committed:
Rename 'WriteHelper' to 'WriteImpl'; refactor 'WriteLine' and 'WriteLineToConsole'

daxian-dbw commented:
@iSazonov I left 3 comments, but pushed 2 commits to address them and do some tiny refactoring. The second commit's message describes itself:
`Rename 'WriteHelper' to 'WriteImpl'; refactor 'WriteLine' and 'WriteLineToConsole'`

Please take a look to see if the changes make sense. Thanks.

iSazonov commented:
@daxian-dbw Sorry, I missed some fixes again. I only have short periods of time to do something useful and for forced pauses I lose context :-(
Thanks for pushing additional commits!

iSazonov commented:
@daxian-dbw Thanks for review and help!

I do not remember what I planned in the continuation of this work. I will look again :-)

The other day I was playing with a small application and added output to the console - this application catastrophically slowed down (up to 10 times). Now I am concerned that our intentions to replace our code with Console.* can lead to negative consequences.

daxian-dbw commented:
The plan was to replace P/Invoke calls with dotnet Console APIs.

> added output to the console - this application catastrophically slowed down (up to 10 times)

I double it's because of the Console API. It could be the fact of writing to console itself caused this difference, no matter you were using P/Invoke or the managed Console API.

daxian-dbw commented:
> The dotnet API (Readline()) is not symmetric for all platforms.

Let's first focus on replacing the write P/Invoke. We can get back to the read operations later. I would like small PRs :)

> Yes, but there is still quite a thick wrap in dotnet.

PSReadline is refactored to move away from P/Invoke to make it more portable. I think perf impact won't be a concern unless we can prove it's noticeable.

iSazonov commented:
> Let's first focus on replacing the write P/Invoke. We can get back to the read operations later. I would like small PRs :)

My current understanding that we can't even do it for Windows - It seems there is no "input completion by tab" in `Readline()`.
https://source.dot.net/#System.Private.CoreLib/shared/System/IO/TextReader.cs,182

daxian-dbw commented:
`Readline()` is the reading part, right? We can focus on the writing part for now, and leave the reading related code untouched.

iSazonov commented:
I can try it. Until I am sure to turn out to completely remove P/Invokes for writing.
